src/business_logic/transaction/internal_objects.rs

```diff
@@ -166,3 +170,246 @@ impl InternalDeploy {
+            TransactionType::DeployAccount,
+            state.count_actual_storage_changes(),
+            None,
+        )
+        .map_err::<StateError, _>(|_| todo!())?;
+
+        Ok(
+            TransactionExecutionInfo::create_concurrent_stage_execution_info(
+                validate_info,
+                Some(constructor_call_info),
+                actual_resources,
+                Some(TransactionType::DeployAccount),
+            ),
+        )
+    }
+
+    pub fn handle_constructor<S>(
+        &self,
+        contract_class: ContractClass,
+        state: &mut S,
+        general_config: &StarknetGeneralConfig,
+        resources_manager: &mut ExecutionResourcesManager,
+    ) -> Result<CallInfo, ExecutionError>
+    where
+        S: Default + State + StateReader,
+    {
+        let num_constructors = contract_class
+            .entry_points_by_type
+            .get(&EntryPointType::Constructor)
+            .map(Vec::len)
+            .unwrap_or(0);
+
+        match num_constructors {
+            0 => {
+                if !self.constructor_calldata.is_empty() {
+                    todo!()
+                }
+
+                Ok(CallInfo::empty_constructor_call(
+                    self.contract_address.clone(),
+                    Address(Felt::zero()),
+                    Some(self.class_hash),
+                ))
+            }
+            _ => self.run_constructor_entrypoint(state, general_config, resources_manager),
+        }
+    }
+
+    pub fn run_constructor_entrypoint<S>(
+        &self,
+        state: &mut S,
+        general_config: &StarknetGeneralConfig,
+        resources_manager: &mut ExecutionResourcesManager,
+    ) -> Result<CallInfo, ExecutionError>
+    where
+        S: Default + State + StateReader,
+    {
+        let call = ExecutionEntryPoint::new(
+            self.contract_address.clone(),
+            self.constructor_calldata.clone(),
+            CONSTRUCTOR_ENTRY_POINT_SELECTOR.clone(),
+            Address(Felt::zero()),
+            EntryPointType::Constructor,
+            None,
+            None,
+        );
+
+        let call_info = call.execute(
+            state,
+            general_config,
+            resources_manager,
+            &self.get_execution_context(general_config.validate_max_n_steps),
+        )?;
+
+        verify_no_calls_to_other_contracts(&call_info)?;
+        Ok(call_info)
+    }
+
+    pub fn get_execution_context(&self, n_steps: u64) -> TransactionExecutionContext {
+        TransactionExecutionContext::new(
+            self.contract_address.clone(),
+            calculate_deploy_account_transaction_hash(
+                self.version,
+                &self.contract_address,
+                Felt::from_bytes_be(&self.class_hash),
+                &self.constructor_calldata,
+                self.max_fee,
+                self.nonce,
+                self.contract_address_salt.0.clone(),
+                self.chain_id.to_felt().to_u64().unwrap(),
```

Comment:
as far as i remember we had changed the chain id to be treated as a felt

Reply:
Why? Doesn't it have only 3 possible values? I think an enum would be far better in this case, especially when compared to a `Felt` which may allocate memory and forces us to check for invalid values.

`InternalDeploy` uses a `u64`. It's probably better than using a felt, but it still forces us to validate its value every time we use it.

Could you confirm whether I may keep `StarknetChainId` or whether I should change it to either `Felt` or `u64`?

Reply:
It only has three possible values, yes but certain hashing functions required it to be a felt, so we add the method .to_felt to manage it. I think that you may keep it like this for now, as there are other functions that also use it as a `u64`